### PR TITLE
Fix snippets/resolver

### PR DESCRIPTION
**Pull request by Austreelis**
The current files in `snippets/resolver` don't configure a DNS resolver, but something closer to a SOA.

### CODE REVIEW

File: snippets/resolver/configuration.nix [hunk 1]
```diff
@@ -1,39 +0,0 @@
-{ ... }:
```
Comment (Austreelis) [inline snippets/resolver/configuration.nix:1]: This file shouldn't be deleted

File: snippets/resolver/default.nix [hunk 2]
```diff
@@ -1,24 +0,0 @@
-{ lib, deploy-rs, nixpkgs, self, ... } @ args:
```
Comment (Austreelis) [inline snippets/resolver/default.nix:1]: This file shouldn't be deleted

File: snippets/resolver/named.conf [hunk 3]
```diff
@@ -1,28 +1,48 @@
 options {
-  directory "/var/cache/bind";
-  listen-on { any; };
-  listen-on-v6 { any; };
-  pid-file "/run/named/named.pid";
+directory "/var/cache/bind";
+allow-recursion {
+192.168.0.0/24;
+127.0.0.1/32;
```
Comment (Austreelis) [inline snippets/resolver/named.conf:5]: Fix indentation:

```suggestion
  directory "/var/cache/bind";
  allow-recursion {
    192.168.0.0/24;
    127.0.0.1/32;
```

Also, not sure the IPs should be those. `127.0.0.1/32` makes sense, but I think `192.168.0.0/24` should be whatever the subnet the resolver should be accessible on is (which when using docker-compose would be `10.0.10.0/24`, as configured in `docker-compose.yml`). In all cases, `localnets` is probably worth a try and should do the right thing.

File: snippets/resolver/named.conf [hunk 4]
```diff
@@ -1,28 +1,48 @@
...
+print-severity NO;
+print-category NO;
+};
+category default {
+"misc";
+};
+category queries {
+"query";
 };
+};
```
Comment (Austreelis) [inline snippets/resolver/named.conf:48]: why removing all indentation ? I don't see any actual change here.

File: snippets/resolver/Dockerfile [hunk 5]
```diff
@@ -1,15 +1,12 @@
 FROM ubuntu/bind9
 
-# MAINTAINER instruction is deprecated
-LABEL maintainers="Austreelis <dev@austreelis.net>"
+MAINTAINER Habibou Moulaye Zeini <h.moulayezeini@students.ephec.be>
```
Comment (Austreelis) [inline snippets/resolver/Dockerfile:3]: As per the above comment, the `MAINTAINER` instruction is deprecated: https://docs.docker.com/engine/reference/builder/#maintainer-deprecated

File: snippets/resolver/Dockerfile [hunk 6]
```diff
@@ -1,15 +1,12 @@
...
 
-# `named`'s log directory
-RUN mkdir /var/log/named && chown -R bind /var/log/named
+COPY named.conf /etc/bind
 
-# Copy this after the rest to avoid invalidating unmodified layers
-COPY named.conf /etc/bind/named.conf
-COPY *.zone /etc/bind/
+RUN mkdir /var/log/named && -R bind /var/log/named
```
Comment (Austreelis) [inline snippets/resolver/Dockerfile:7]: Creating the log directory here isn't ideal, as previous layers may change and cause this layer to be rebuilt without reason.

File: snippets/resolver/Dockerfile [hunk 7]
```diff
@@ -1,15 +1,12 @@
...
-# MAINTAINER instruction is deprecated
-LABEL maintainers="Austreelis <dev@austreelis.net>"
+MAINTAINER Habibou Moulaye Zeini <h.moulayezeini@students.ephec.be>
 
-# `named`'s log directory
-RUN mkdir /var/log/named && chown -R bind /var/log/named
+COPY named.conf /etc/bind
 
-# Copy this after the rest to avoid invalidating unmodified layers
-COPY named.conf /etc/bind/named.conf
```
Comment (Austreelis) [inline snippets/resolver/Dockerfile:10]: Any reason to change this `COPY` instruction ? There is no real difference as far as I can tell.

File: snippets/resolver/Dockerfile [hunk 8]
```diff
@@ -1,15 +1,12 @@
...
-COPY *.zone /etc/bind/
+RUN mkdir /var/log/named && -R bind /var/log/named
 
-# Always start bind's resolver
-#ENTRYPOINT ["/usr/local/bin/named" "-g"]
+EXPOSE 53/tcp
+EXPOSE 53/udp
 
+CMD ["named", "-g"]
```
Comment (Austreelis) [inline snippets/resolver/Dockerfile:12]: The `ubuntu/bind9` image already configures the right `ENTRYPOINT` and `CMD` instruction, so there shouldn't be a need to do it here.

File: snippets/resolver/Dockerfile [hunk 9]
```diff
@@ -1,15 +1,12 @@
 FROM ubuntu/bind9
 
-# MAINTAINER instruction is deprecated
-LABEL maintainers="Austreelis <dev@austreelis.net>"
+MAINTAINER Habibou Moulaye Zeini <h.moulayezeini@students.ephec.be>
 
-# `named`'s log directory
-RUN mkdir /var/log/named && chown -R bind /var/log/named
+COPY named.conf /etc/bind
```
Comment (Austreelis) [inline snippets/resolver/Dockerfile:5]: `COPY`ing a config file subject to change *first* will invalidate all following layers whenever it is updated and force a full rebuild of all the next layers. I get that it's no big deal given the number of layers this image adds on `ubuntu/bind9`, but it still isn't ideal imho.

File: snippets/resolver/Dockerfile [hunk 10]
```diff
@@ -1,15 +1,12 @@
...
 
-# Copy this after the rest to avoid invalidating unmodified layers
-COPY named.conf /etc/bind/named.conf
-COPY *.zone /etc/bind/
+RUN mkdir /var/log/named && -R bind /var/log/named
 
-# Always start bind's resolver
-#ENTRYPOINT ["/usr/local/bin/named" "-g"]
+EXPOSE 53/tcp
+EXPOSE 53/udp
```
Comment (Austreelis) [inline snippets/resolver/Dockerfile:10]: The ubuntu/bind9 image already exposes those ports, so there shouldn't be a need to do it here.

File: snippets/resolver/named.conf [hunk 11]
```diff
@@ -1,28 +1,48 @@
...
+// Configure the IPs to listen on here.
+listen-on port 53 {any;};
+listen-on-v6 { none; };
+pid-file "/var/run/named/named.pid";
+allow-transfer { none; };
+};
+zone "." IN {
+type hint;
+file "/usr/share/dns/root.hints";
+};
```
Comment (Austreelis) [inline snippets/resolver/named.conf:16]: Fix indentation:

```suggestion
zone "." IN {
  type hint;
  file "/usr/share/dns/root.hints";
};

```

Also, that zone file isn't included in the docker image, which is probably one of the reason why name resolution doesn't work for generic addresses.

File: snippets/resolver/named.conf [hunk 12]
```diff
@@ -1,28 +1,48 @@
...
+zone "localhost" IN {
+type master;
+file "db.local";
+allow-update { none; };
+notify no;
+};
```
Comment (Austreelis) [inline snippets/resolver/named.conf:22]: Fix indentation:

```suggestion
zone "localhost" IN {
  type master;
  file "db.local";
  allow-update { none; };
  notify no;
};

```

Also, that db file isn't included in the docker image. I'm also not sure we need such a zone for a resolver (even a local one).

File: snippets/resolver/named.conf [hunk 13]
```diff
@@ -1,28 +1,48 @@
...
+zone "127.in-addr.arpa" IN {
+type master;
+file "db.127";
+allow-update { none; };
+notify no;
```
Comment (Austreelis) [inline snippets/resolver/named.conf:27]: Fix indentation:

```suggestion
zone "127.in-addr.arpa" IN {
  type master;
  file "db.127";
  allow-update { none; };
  notify no;
```

Same as the previous 2, that db file isn't included in the docker image. Also, I'm not sure we need such a reverse mapping for a resolver ?

File: snippets/resolver/named.conf [hunk 14]
```diff
@@ -1,28 +1,48 @@
...
+192.168.0.0/24;
+127.0.0.1/32;
 };
-
-zone "2tm1-admin-sys2.ephec.austreelis.net" IN {
-  type master;
-  file "/etc/bind/2tm1-admin-sys2.ephec.austreelis.net.zone";
+// Configure the IPs to listen on here.
+listen-on port 53 {any;};
+listen-on-v6 { none; };
```
Comment (Austreelis) [inline snippets/resolver/named.conf:9]: Fix indentation. Also, I'm pretty sure we don't want to opt *out* of IPv6 (even if we don't use it for now). Enabling IPv6 should probably be done on the network/routing level, and we should let `named` use it *if available* in all cases.

```suggestion
  listen-on-v6 { any; };
```

File: snippets/resolver/named.conf [hunk 15]
```diff
@@ -1,28 +1,48 @@
...
+allow-recursion {
+192.168.0.0/24;
+127.0.0.1/32;
 };
-
-zone "2tm1-admin-sys2.ephec.austreelis.net" IN {
-  type master;
-  file "/etc/bind/2tm1-admin-sys2.ephec.austreelis.net.zone";
+// Configure the IPs to listen on here.
+listen-on port 53 {any;};
```
Comment (Austreelis) [inline snippets/resolver/named.conf:8]: Fix indentation:

```suggestion
   // Configure the IPs to listen on here.
  listen-on port 53 {any;};
```

File: snippets/resolver/named.conf [hunk 16]
```diff
@@ -1,28 +1,48 @@
...
+listen-on-v6 { none; };
+pid-file "/var/run/named/named.pid";
+allow-transfer { none; };
+};
```
Comment (Austreelis) [inline snippets/resolver/named.conf:12]: Fix indentation:
```suggestion
  pid-file "/var/run/named/named.pid";
  allow-transfer { none; };
};

```

File: snippets/resolver/named.conf [hunk 17]
```diff
@@ -1,28 +1,48 @@
...
+allow-update { none; };
+notify no;
+};
+zone "127.in-addr.arpa" IN {
+type master;
+file "db.127";
+allow-update { none; };
+notify no;
 };
-
```
Comment (Austreelis) [inline snippets/resolver/named.conf:12]: Removing empty lines like this one makes the file less readable imo.

File: snippets/resolver/Dockerfile (hunk 7 above)
Comment (hmoulayez) [reply snippets/resolver/Dockerfile:10]: Not really. There is no difference with the previous one

File: snippets/resolver/Dockerfile (hunk 6 above)
Comment (hmoulayez) [reply snippets/resolver/Dockerfile:7]: Ok

File: snippets/resolver/Dockerfile (hunk 10 above)
Comment (hmoulayez) [reply snippets/resolver/Dockerfile:10]: Ack

File: snippets/resolver/Dockerfile (hunk 8 above)
Comment (hmoulayez) [reply snippets/resolver/Dockerfile:12]: Ack

File: snippets/resolver/named.conf (hunk 3 above)
Comment (hmoulayez) [reply snippets/resolver/named.conf:5]: Ack

File: snippets/resolver/named.conf (hunk 16 above)
Comment (hmoulayez) [reply snippets/resolver/named.conf:12]: Ack

File: snippets/resolver/named.conf (hunk 11 above)
Comment (hmoulayez) [reply snippets/resolver/named.conf:17]: I don't know how it is possible but this file is present in my docker image. So, I think the problem is not there

File: snippets/resolver/named.conf (hunk 12 above)
Comment (hmoulayez) [reply snippets/resolver/named.conf]: The file comes with the bind package. I checked and it is present.  I think the role of such is not is to link the hostname "localhost" to the loopback addresses "127.0.0.1" and "::1"

File: snippets/resolver/named.conf (hunk 13 above)
Comment (hmoulayez) [reply snippets/resolver/named.conf]: Same as the previous

File: snippets/resolver/named.conf (hunk 4 above)
Comment (hmoulayez) [reply snippets/resolver/named.conf]: Not intentionnal. It is due to my text editor. Will pay more attention next time

File: snippets/resolver/Dockerfile (hunk 9 above)
Comment (hmoulayez) [reply snippets/resolver/Dockerfile:5]: Ack

**Comment by Austreelis** [comment]
I pushed a few commits (f413e7b to c69afc3). Most changes are pretty trivials and shouldn't be controversial. But I did set `allow-recursion` in `named.conf` to `localhost` & `localnets`. This still works (as expected), and I think should be preferred, as it removes the need to keep the allowed addresses up-to-date with the network configuration. I.e. if you don't want something to be able to query the resolver, don't connect it to the resolver's (virtual) network.

This setting (`allow-recursion { localhost; localnets; };`) is actually the default, so we could just not specify it. I think it's good to be explicit though.

Note: The `localnets` network filter is explained [here](https://bind9.readthedocs.io/en/v9_16_6/reference.html#acl-statement-definition-and-usage). TL;DR: It matches any host on an IPv4 or IPv6 network for which the system has an interface, which in our case would be any host connected to the `resolver-nw` docker network configured in `docker-compose.yml`.

**Review by hmoulayez — APPROVED** [review]
The Docker and named.conf files are OK for me. 
Didn't ckecked the .nix files.

**Pull request merged**